JSDK-2338: add region in connectOptions

makarandp0 opened the pull request:
@manjeshbhargav 

This change adds a `region` to connectOptions. The  `region` specified would be used for connecting to signaling server.  

TODO:
- [x] make and verify region change
- [x] add unit tests
- [x] add integration tests
- [x] fix the link for available regions  
- [x] add region option in the signaling-sample app [https://code.hq.twilio.com/client/simpler-signaling.appspot.com/pull/33]

manjeshbhargav commented on lib/connect.js:
```diff
@@ -303,6 +304,9 @@ function connect(token, options) {
  *   for the {@link LocalParticipant} to {@link NetworkQualityVerbosity}<code style="padding:0 0">#minimal</code>
  *   and the {@link NetworkQualityVerbosity} for {@link RemoteParticipant}s to
  *   {@link NetworkQualityVerbosity}<code style="padding:0 0">#none</code>.
+ * @property {?string} [region='gll'] - Preferred region. This region will be used
```
@makarandp0 Two things here:
1. Can you convert `{?string} to `{string}`? A "?" implies that the argument is nullable.

2. Can we make the description closer to the spec:
```
- Preferred region; By default, the client will connect to the nearest signaling server
determined by latency based routing. Setting a value other than <code style="padding:0 0">gll</code>
bypasses routing and guarantees that signaling traffic will be terminated in the region that you prefer;
If you are connecting to a Group Room created with the <code style="padding:0 0">gll</code>
Media Region (either Ad-Hoc or via the REST API), then the Room's Media Region will be selected
based upon your Client's region
```

makarandp0 replied:
Thanks. makes sense for both.

manjeshbhargav commented on test/unit/spec/util/constants.js:
```diff
@@ -0,0 +1,30 @@
+const Constants = require('../../../../lib/util/constants');
+
+describe('Constants', () => {
+  describe('WS_SERVER.', () => {
```
@makarandp0 

IMHO, these tests can be moved to the `connect` unit tests:

*test/unit/spec/connect.js*
```js
describe('connect', () => {
  ...
  describe('called with ConnectOptions#region', () => {
    [/* Values for region */].forEach(region => {
      context(`When region is "${region}"`, () => {
        ...
      });
    });
  });

  describe('called with ConnectOptions#environment', () => {
    [/* Values for environment */].forEach(environment => {
      context(`When environment is "${environment}"`, () => {
        ...
      });
    });
  });
  ...
});
```

makarandp0 replied:
I did add a test to connect to make sure that integration between connect and util works. But these unit tests, I think are more useful to be closer to implementation.

makarandp0 replied:
now I have added similar tests to connct.js as well.

manjeshbhargav replied:
@makarandp0 

Since `utils` is mostly used by other modules, we should test them by actually testing the context in which they will be used. That way, when anyone reads the tests, they know that ConnectOptions#region is expected to behave a certain way.

makarandp0 replied:
@manjeshbhargav,  I did take your suggestion and added tests in connect.js My reasoning for keeping the tests in constants.WS_SERVER was the ability to go into details of how the format of the url  looks like for example:

```
    it('maps gll to global for region', () => {
      const wsServer = Constants.WS_SERVER('anyenvironment', 'gll');
      assert.equal(wsServer, 'wss://global.vss.anyenvironment.twilio.com/signaling');
    });
```
where as similar test when written from connect.js need to  be bit vague about the actual format as they do not need to know what Constants.WS_SERVER does interenally. So I had something like:
```
        if (region === 'gll') {
          assert(options.wsServer.indexOf('global') > 0);
        } else {
          assert(options.wsServer.indexOf(region) > 0);
        }
```

makarandp0 commented on CHANGELOG.md:
```diff
@@ -5,6 +5,19 @@ For 1.x changes, go [here](https://github.com/twilio/twilio-video.js/blob/suppor
 
 New Features
 ------------
+- [When connecting to a room](https://www.twilio.com/docs/video/javascript-v2-getting-started#connect-to-a-room),
```
thanks, that reads much better :) I will fix.

makarandp0 commented on CHANGELOG.md:
````diff
@@ -5,6 +5,19 @@ For 1.x changes, go [here](https://github.com/twilio/twilio-video.js/blob/suppor
+    region: 'de1'
+  });
+  ```
+  Above option will use signaling servers in Germany. For other possible values for region
````
Thanks @manjeshbhargav . I will fix this the text. 
About the link - it mentions more regions than what we have currently ( of those regions only us1, in1 and de1 work right now) - @innerverse - where should we point to for definitive list of regions that developers can use?

innerverse replied:
@makarandp0 - We will deploy to every region listed once any SDK goes GA.

makarandp0 commented:
Thanks you for your feedback @manjeshbhargav. I have incorporated it and also added integration tests. Can you take another look when you get a chance. Thanks!

manjeshbhargav commented on test/integration/spec/connect.js:
```diff
@@ -103,6 +104,52 @@ describe('connect', function() {
     });
   });
 
+  describe('region selection ', async () => {
```
@makarandp0 Nitpick: Remove extra space after `selection`?

manjeshbhargav commented on test/integration/spec/connect.js:
```diff
@@ -103,6 +104,52 @@ describe('connect', function() {
+      token = getToken(identity);
+    });
+
+    context('with an invalid region', async () => {
```
@makarandp0 You can simplify the cases here and re-use some of the test code:
```js
['without', 'global', 'specific', 'invalid'].forEach(regionType => {
  const regionOptions = {
    global: { region: 'gll' },
    invalid: { region: 'foo' },
    specific: { region: 'de1' },
    without: {}
  }[regionType];

  context(`when called ${regionType === 'without a' ? regionType : `with a(n) ${regionType}`} region`, () => {
    beforeEach(() => {
      cancelablePromise = connect(token, Object.assign({}, defaults, regionOptions, { tracks: [] }));
    });

    if (regionType === 'invalid') {
      it('should return a CancelablePromise that rejects with a SignalingConnectionError', async () => {
        ...
      });
    } else {
      it('should return a CancelablePromise that resolves with a Room', async () => {
        ...
      });
    }
  });
});
```

makarandp0 replied:
Thanks for the suggestion, that does look more compact.

manjeshbhargav commented on test/integration/spec/connect.js:
```diff
@@ -103,6 +104,52 @@ describe('connect', function() {
+        try {
+          const room = await cancelablePromise;
+          room.disconnect();
+          throw new Error(`Connected to ${room.sid} with an invalid LogLevel`);
```
@makarandp0 "with an invalid ~~LogLevel~~ region"?

manjeshbhargav commented on test/integration/spec/connect.js:
```diff
@@ -103,6 +104,52 @@ describe('connect', function() {
+          throw new Error(`Connected to ${room.sid} with an invalid LogLevel`);
+        } catch (error) {
+          assert(error instanceof SignalingConnectionError);
+          assert(/Signaling connection error/.test(error.message));
```
@makarandp0 Testing error message is not necessary.

makarandp0 commented on test/unit/spec/connect.js:
```diff
@@ -36,17 +36,66 @@ describe('connect', () => {
+        });
+
+        const options = signaling.args[0][1];
+        assert.equal(options.wsServer, WS_SERVER(options.environment, options.region));
```
I was debating about this when I added tests for `constants.WS_SERVER` in separate file. One option (a) was to let tests in connect.js know about implementation of WS_SERVER (like you are suggesting) or b)  add specific tests into constants.WS_SERVER. 
I chose b) but I am don't feel very strongly about it. I will delete the constants.js tests and add  the logic here.

manjeshbhargav commented on CHANGELOG.md:
````diff
@@ -5,6 +5,19 @@ For 1.x changes, go [here](https://github.com/twilio/twilio-video.js/blob/suppor
+  const room = await connect(token, {
+    region: 'de1'
+  });
+  ```
````
@makarandp0 Nitpick: Can you add a new line here?

makarandp0 replied:
@aymenn can you take a look at this changelog entry please, and suggest any updates.

ceaglest replied:
Suggestion: "terminate at the" -> "terminate in the"

Overall, this sounds good to me, specifically leaving out mention of "gll" media behavior which is not ready yet. We will need to update the white listing guide to mention the us1, de1, and in1 VSS deployments sometime soon.

makarandp0 replied:
updated per your suggestion @ceaglest.
